### PR TITLE
Add OpenGL interop

### PR DESCRIPTION
**Pull request by neon60**

### CODE REVIEW

File: docs/how-to/hip_runtime_api/opengl_interop.rst [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                 OpenGL's graphics rendering.
+   :keywords: AMD, ROCm, HIP, OpenGL, interop, interoperability
+
+********************************************************************************
+OpenGL interoperability
+********************************************************************************
+
+The HIP--OpenGL interoperation involves mapping OpenGL resources, such as
+buffers and textures, for HIP access. This mapping process enables HIP to
+utilize these resources directly, bypassing the need for data transfers between
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/opengl_interop.rst]:
```suggestion
utilize these resources directly, bypassing the need for costly data transfers between
```

File: docs/how-to/hip_runtime_api/opengl_interop.rst [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+the CPU and GPU. This capability is useful in applications that require both
+intensive GPU computation and real-time visualization.
+
+The OpenGL resource can be mapped to HIP with :cpp:func:`hipGraphicsMapResources`
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/opengl_interop.rst]: I feel like these functions deserve a little more description as to how the user should use them.  For instance, the reference material provides little explanation about the various parameters to the function. What does "Stream for synchronization" refer to in the hipGraphicsMapResources() function? How does one determine the "Size of the buffer accessible from devPtr" for the hipGraphicsResourceGetMappedPointer() function?

File: docs/how-to/hip_runtime_api/opengl_interop.rst [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+where a simple HIP kernel is used to simulate a sine wave and rendered to a
+window as a grid of triangles using OpenGL. For a working example, there are
+multiple initialization steps needed like creating and opening a window, 
+initializing OpenGL or selecting the OpenGL-capable device. After the initialization
+in the example, the kernel simulates the sinewave and updates the window's
+framebuffer in a cycle until the window is not closed.
+
+.. note::
+
+   The more recent recent OpenGL functions loaded with `OpenGL loader <https://github.com/ROCm/rocm-examples/tree/develop/External/glad>`_,
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/opengl_interop.rst]: This sentence seems to be missing something, like "the more recent functions _are_ loaded"?

File: docs/how-to/hip_runtime_api/opengl_interop.rst [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+   as these are not loaded by default on all platforms. The use of custom loader
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/opengl_interop.rst]:
```suggestion
   as these are not loaded by default on all platforms. The use of a custom loader
```

File: docs/how-to/hip_runtime_api/opengl_interop.rst [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+   shown in the following example
+
+   .. literalinclude:: ../../tools/example_codes/opengl_interop.hip
+      :start-after: // [Sphinx opengl functions load start]
+      :end-before: // [Sphinx opengl functions load end]
+
+The OpenGL buffer is imported to HIP using as the following way:
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/opengl_interop.rst]:
```suggestion
The OpenGL buffer is imported to HIP in the following way:
```

File: docs/how-to/hip_runtime_api/opengl_interop.rst [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+.. literalinclude:: ../../tools/example_codes/opengl_interop.hip
+   :start-after: // [Sphinx buffer register and get start]
+   :end-before: // [Sphinx buffer register and get end]
+
+The imported pointer manipulated in the sinewave kernel as the following way:
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/opengl_interop.rst]:
```suggestion
The imported pointer is manipulated in the sinewave kernel as shown in the following example:
```

File: docs/how-to/hip_runtime_api/opengl_interop.rst [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+.. literalinclude:: ../../tools/example_codes/opengl_interop.hip
+   :start-after: /// [Sphinx sinewave kernel start]
+   :end-before: /// [Sphinx sinewave kernel end]
+
+.. literalinclude:: ../../tools/example_codes/opengl_interop.hip
+   :start-after: // [Sphinx buffer use in kernel start]
+   :end-before: // [Sphinx buffer use in kernel end]
+
+The OpenGL buffer is imported to HIP using as the following way:
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/opengl_interop.rst]:
```suggestion
The OpenGL buffer is imported to HIP in the following way:
```

**Review by randyh62 — APPROVED** [review]
In general I feel like this How to topic could use a little more description of how to use the functions.

File: docs/how-to/hip_runtime_api/opengl_interop.rst (hunk 3 above)
Comment (neon60) [reply docs/how-to/hip_runtime_api/opengl_interop.rst]: Updated, please check.

File: docs/how-to/hip_runtime_api/opengl_interop.rst (hunk 2 above)
Comment (neon60) [reply docs/how-to/hip_runtime_api/opengl_interop.rst]: I updated the description here, but I would not add the argument description. "Stream for synchronization" should be described in the section where the stream is introduced while the size of the buffer should be known by the developer, before using the map function.

File: docs/how-to/hip_runtime_api/opengl_interop.rst [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+multiple initialization steps needed like creating and opening a window, 
+initializing OpenGL or selecting the OpenGL-capable device. After the initialization
+in the example, the kernel simulates the sinewave and updates the window's
+framebuffer in a cycle until the window is not closed.
+
+.. note::
+
+   The more recent OpenGL functions are loaded with `OpenGL loader <https://github.com/ROCm/rocm-examples/tree/develop/External/glad>`_,
+   as these are not loaded by default on all platforms. The use of a custom
+   loader shown in the following example
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/opengl_interop.rst]:
```suggestion
   loader is shown in the following example
```

File: docs/how-to/hip_runtime_api/opengl_interop.rst (hunk 2 above)
Comment (randyh62) [reply docs/how-to/hip_runtime_api/opengl_interop.rst]: Is the size of the buffer static, or varies depending on what the developer is trying to do? If the size is static according to the mapped buffer, then why does the user need to specify it? If the size of the buffer depends on the users needs, then how do they ensure the requested size does not exceed the mapped size?

Comment (neon60) [reply docs/how-to/hip_runtime_api/opengl_interop.rst]: The OpenGL buffer size not static, and I checked we can get the buffer size with the following function call:

```
GLint size = 0;

glGetBufferParameteriv(GL_ARRAY_BUFFER, GL_BUFFER_SIZE, &size);
```

The hipGraphicsMapResources function returns error codes (hipSuccess, hipErrorInvalidValue, hipErrorUnknown or hipErrorInvalidResourceHandle), but I am not sure, if it's checking the buffer size. If the size of the buffer invalid, we may get a HIP error during the mapping, or we will get into trouble, while we try to use the mapped resource. In C and C++ the invalid memory access can happen.

"I feel like these functions deserve a little more description"

I wanted to add more here, but simple there is not much more I can write about this function. I can add a note for glGetBufferParameteriv function, if it helps.

Comment (randyh62) [reply docs/how-to/hip_runtime_api/opengl_interop.rst]: Thanks for checking. I think I will touch base with Julia as some of this information should appear in the API reference.

File: docs/how-to/hip_runtime_api/opengl_interop.rst [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+                 efficient data sharing between HIP's computing power and
+                 OpenGL's graphics rendering.
+   :keywords: AMD, ROCm, HIP, OpenGL, interop, interoperability
+
+*******************************************************************************
+OpenGL interoperability
+*******************************************************************************
+
+The HIP--OpenGL interoperation involves mapping OpenGL resources, such as
+buffers and textures, for HIP access. This mapping process enables HIP to
```
Comment (jujiang-del) [inline docs/how-to/hip_runtime_api/opengl_interop.rst]: For HIP to interact with OpenGL.

File: docs/how-to/hip_runtime_api/opengl_interop.rst [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+Example
+===============================================================================
+
+ROCm examples have a `HIP--OpenGL interoperation example <https://github.com/ROCm/rocm-examples/tree/develop/HIP-Basic/opengl_interop>`_,
+where a simple HIP kernel is used to simulate a sine wave and rendered to a
+window as a grid of triangles using OpenGL. For a working example, there are
+multiple initialization steps needed like creating and opening a window, 
+initializing OpenGL or selecting the OpenGL-capable device. After the
+initialization in the example, the kernel simulates the sinewave and updates
+the window's framebuffer in a cycle until the window is not closed.
```
Comment (jujiang-del) [inline docs/how-to/hip_runtime_api/opengl_interop.rst]: until the windows is closed.

File: docs/how-to/hip_runtime_api/opengl_interop.rst [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+Unmapping resources with :cpp:func:`hipGraphicsUnmapResources` after
+computations ensure proper resource management.
+
+Example
+===============================================================================
+
+ROCm examples have a `HIP--OpenGL interoperation example <https://github.com/ROCm/rocm-examples/tree/develop/HIP-Basic/opengl_interop>`_,
+where a simple HIP kernel is used to simulate a sine wave and rendered to a
+window as a grid of triangles using OpenGL. For a working example, there are
```
Comment (jujiang-del) [inline docs/how-to/hip_runtime_api/opengl_interop.rst:35]: To communicate with OpenGL, there are multiple...

File: docs/how-to/hip_runtime_api/opengl_interop.rst [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+   :language: cpp
+
+.. literalinclude:: ../../tools/example_codes/opengl_interop.hip
+   :start-after: // [Sphinx buffer use in kernel start]
+   :end-before: // [Sphinx buffer use in kernel end]
+   :language: cpp
+
+.. <!-- spellcheck-enable -->
+
+The HIP graphics resource that is imported from the OpenGL buffer has to be
```
Comment (jujiang-del) [inline docs/how-to/hip_runtime_api/opengl_interop.rst]: should be unmapped/unregistered...

File: docs/how-to/hip_runtime_api/opengl_interop.rst (hunk 9 above)
Comment (neon60) [reply docs/how-to/hip_runtime_api/opengl_interop.rst]: Updated.

File: docs/how-to/hip_runtime_api/opengl_interop.rst (hunk 10 above)
Comment (neon60) [reply docs/how-to/hip_runtime_api/opengl_interop.rst]: Updated the text.

File: docs/how-to/hip_runtime_api/opengl_interop.rst (hunk 11 above)
Comment (neon60) [reply docs/how-to/hip_runtime_api/opengl_interop.rst:35]: The OpenGL application needs initializations steps, that's why I highlighted here that a working example is longer.

The suggestion would change the meaning of the sentence incorrectly. These initialization steps needs for OpenGL application, not for the communication with OpenGL.

If you want, I can do: "For a working HIP--OpenGL interoperation example, there are"

File: docs/how-to/hip_runtime_api/opengl_interop.rst (hunk 12 above)
Comment (neon60) [reply docs/how-to/hip_runtime_api/opengl_interop.rst]: Updated the text, please check now.

File: docs/how-to/hip_runtime_api/opengl_interop.rst (hunk 11 above)
Comment (jujiang-del) [reply docs/how-to/hip_runtime_api/opengl_interop.rst:35]: sure.

**Review by jujiang-del — APPROVED** [review]

**Pull request merged**